File: screenshot/Pages/Tutorial/TutorialViewController.m
//
//  TutorialViewController.m
//  screenshot
//
//  Created by Corey Werner on 8/9/17.
//  Copyright © 2017 crazeapp. All rights reserved.
//

#import "TutorialViewController.h"
#import "TutorialEmailSlideView.h"
#import "TutorialWelcomeSlideView.h"
#import "TutorialTrySlideView.h"
#import "Geometry.h"
#import "PermissionsManager.h"
#import "WebViewController.h"
#import "AnalyticsManager.h"
#import "screenshot-Swift.h"

@interface TutorialViewController () <UIScrollViewDelegate, TutorialWelcomeSlideViewDelegate, TutorialEmailSlideViewDelegate, TutorialTrySlideViewDelegate> {
    BOOL _didPresentDeterminePushAlertController;
    BOOL _scrollViewIsScrollingAnimation;
}

@property (nonatomic, strong) UIScrollView *scrollView;
@property (nonatomic, strong) UIView *contentView;
@property (nonatomic, strong) NSArray<TutorialBaseSlideView *>* slides;
@property (nonatomic, strong) UpdatePromptHandler *handler;

@end

@implementation TutorialViewController

#pragma mark - Life Cycle

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        self.title = @"Tutorial";
        
        [AnalyticsManager track:@"Started Tutorial"];
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.handler = [[UpdatePromptHandler alloc] initWithContainerViewController:self];
    [self.handler start];
    
    self.view.backgroundColor = [UIColor whiteColor];
    
    self.scrollView = ({
        UIScrollView *scrollView = [[UIScrollView alloc] init];
        scrollView.translatesAutoresizingMaskIntoConstraints = NO;
        scrollView.delegate = self;
        scrollView.showsHorizontalScrollIndicator = NO;
        scrollView.pagingEnabled = YES;
        scrollView.scrollEnabled = NO;
        [self.view addSubview:scrollView];
        [scrollView.topAnchor constraintEqualToAnchor:self.topLayoutGuide.bottomAnchor].active = YES;
        [scrollView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor].active = YES;
        [scrollView.bottomAnchor constraintEqualToAnchor:self.bottomLayoutGuide.topAnchor].active = YES;
        [scrollView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor].active = YES;
        scrollView;
    });
    
    UIView *contentView = ({
        UIView *view = self.contentView;
        view.translatesAutoresizingMaskIntoConstraints = NO;
        [self.scrollView addSubview:view];
        [view.topAnchor constraintEqualToAnchor:self.topLayoutGuide.bottomAnchor].active = YES;
        [view.leadingAnchor constraintEqualToAnchor:self.scrollView.leadingAnchor].active = YES;
        [view.bottomAnchor constraintEqualToAnchor:self.bottomLayoutGuide.topAnchor].active = YES;
        [view.trailingAnchor constraintEqualToAnchor:self.scrollView.trailingAnchor].active = YES;
        [view.widthAnchor constraintEqualToAnchor:self.scrollView.widthAnchor multiplier:self.slides.count].active = YES;
        [view.heightAnchor constraintEqualToAnchor:self.scrollView.heightAnchor].active = YES;
        view;
    });
    
    CGFloat p = [Geometry padding];
    CGFloat t = self.contentView.layoutMargins.top;
    
    for (NSInteger i = 0; i < self.slides.count; i++) {
        TutorialBaseSlideView *slide = self.slides[i];
        slide.translatesAutoresizingMaskIntoConstraints = NO;
        [contentView addSubview:slide];
        slide.layoutMargins = UIEdgeInsetsMake(p, p, p, p);
        [slide.widthAnchor constraintEqualToAnchor:self.scrollView.widthAnchor].active = YES;
        [slide.heightAnchor constraintEqualToAnchor:self.scrollView.heightAnchor constant:-t].active = YES;
        [slide.topAnchor constraintEqualToAnchor:contentView.topAnchor constant:t].active = YES;
        
        if (i == 0) {
            [slide.leadingAnchor constraintEqualToAnchor:contentView.leadingAnchor].active = YES;
            [slide didEnterSlide];
            
        } else if (i == self.slides.count - 1) {
            [slide.trailingAnchor constraintEqualToAnchor:contentView.trailingAnchor].active = YES;
        }
        
        if (i > 0) {
            TutorialBaseSlideView *previousSlide = self.slides[i - 1];
            [slide.leadingAnchor constraintEqualToAnchor:previousSlide.trailingAnchor].active = YES;
        }
    }
}

- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator {
    [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
    
    NSUInteger currentSlideIndex = [self currentSlideIndex];
    
    [coordinator animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext>  _Nonnull context) {
        CGPoint offset = self.scrollView.contentOffset;
        offset.x = size.width * currentSlideIndex;
        self.scrollView.contentOffset = offset;
    } completion:nil];
}


#pragma mark - Layout

- (UIView *)contentView {
    if (!_contentView) {
        _contentView = [[UIView alloc] init];
    }
    return _contentView;
}

- (void)setContentLayoutMargins:(UIEdgeInsets)contentLayoutMargins {
    self.contentView.layoutMargins = contentLayoutMargins;
}

- (UIEdgeInsets)contentLayoutMargins {
    return self.contentView.layoutMargins;
}


#pragma mark - Slides

- (NSArray<TutorialBaseSlideView *> *)slides {
    if (!_slides) {
        TutorialWelcomeSlideView *welcomeSlideView = [[TutorialWelcomeSlideView alloc] init];
        welcomeSlideView.delegate = self;
        
        TutorialEmailSlideView *emailSlideView = [[TutorialEmailSlideView alloc] init];
        emailSlideView.delegate = self;
        
        TutorialTrySlideView *trySlideView = [[TutorialTrySlideView alloc] init];
        trySlideView.delegate = self;
        
        _slides = @[welcomeSlideView,
                    emailSlideView,
                    trySlideView
                    ];
    }
    return _slides;
}

- (NSUInteger)currentSlideIndex {
    return ceil(self.scrollView.contentOffset.x / self.scrollView.bounds.size.width);
}

- (TutorialBaseSlideView *)currentSlide {
    return self.slides[[self currentSlideIndex]];
}

- (void)tutorialWelcomeSlideViewDidComplete:(TutorialWelcomeSlideView *)slideView {
    slideView.delegate = nil;
    
    if ([[NSUserDefaults standardUserDefaults] boolForKey:UserDefaultsKeys.tutorialCompleted]) {
        // The tutorial is being presented elsewhere and shouldn't
        // include the try slide
        
        [self.delegate tutorialViewControllerDidComplete:self];
        
    } else {
        [self scrollToNextSlide];
    }
}

- (void)tutorialEmailSlideViewDidComplete:(TutorialEmailSlideView *)slideView {
    slideView.delegate = nil;
    [self scrollToNextSlide];
}

- (void)tutorialEmailSlideViewDidTapTermsOfService:(TutorialEmailSlideView *)slideView {
    UIViewController *viewController = [TutorialEmailSlideView termsOfServiceViewControllerWithDoneTarget:self doneAction:@selector(dismissViewController)];
    [self presentViewController:viewController animated:YES completion:nil];
}

- (void)tutorialEmailSlideViewDidTapPrivacyPolicy:(TutorialEmailSlideView *)slideView {
    UIViewController *viewController = [TutorialEmailSlideView privacyPolicyViewControllerWithDoneTarget:self doneAction:@selector(dismissViewController)];
    [self presentViewController:viewController animated:YES completion:nil];
}

- (void)tutorialTrySlideViewDidComplete:(TutorialTrySlideView *)slideView {
    slideView.delegate = nil;
    
    [[NSUserDefaults standardUserDefaults] setBool:YES forKey:UserDefaultsKeys.tutorialCompleted];
    
    [self.delegate tutorialViewControllerDidComplete:self];
    [AnalyticsManager track:@"Finished Tutorial"];
}


#pragma mark - Scroll View

- (void)scrollToNextSlide {
    if (!_scrollViewIsScrollingAnimation) {
        _scrollViewIsScrollingAnimation = YES;
        
        [[self currentSlide] willLeaveSlide];
        
        CGPoint offset = CGPointZero;
        offset.x = self.scrollView.bounds.size.width + self.scrollView.contentOffset.x;
        
        [self.scrollView setContentOffset:offset animated:YES];
    }
}

- (void)scrollViewDidEndScrollingAnimation:(UIScrollView *)scrollView {
    _scrollViewIsScrollingAnimation = NO;
    
    [[self currentSlide] didEnterSlide];
}


#pragma mark - Navigation

- (void)dismissViewController {
    [self dismissViewControllerAnimated:YES completion:nil];
}

@end

File: screenshot/Pages/Tutorial/TutorialViewController.swift
//
//  TutorialViewController.swift
//  screenshot
//
//  Created by Jacob Relkin on 10/14/17.
//  Copyright © 2017 crazeapp. All rights reserved.
//

import UIKit

@objc protocol TutorialViewControllerDelegate : class {
    func tutoriaViewControllerDidComplete(_ viewController: TutorialViewController)
}

class TutorialViewController : UIViewController {
    weak var delegate: TutorialViewControllerDelegate?
    
    var video = TutorialVideo.Standard
    var updatePromptHandler: UpdatePromptHandler!
    let scrollView = UIScrollView()
    let contentView = UIView()
    
    var contentLayoutMargins: UIEdgeInsets {
        get {
            return contentView.layoutMargins
        } set {
            contentView.layoutMargins = newValue
        }
    }
    
    private var didPresentDeterminePushAlertController:Bool = false
    fileprivate var scrollViewIsScrollingAnimation:Bool = false
    
    // MARK: Slides
    
    var slides = [TutorialBaseSlideView]()
    
    private func buildSlides() -> [TutorialBaseSlideView] {
        let welcomeSlide = TutorialWelcomeSlideView()
        welcomeSlide.getStartedButtonTapped = presentVideo
        
        let emailSlide = TutorialEmailSlideView()
        emailSlide.delegate = self
        
        let trySlide = TutorialTrySlideView()
        trySlide.delegate = self
        
        return [welcomeSlide, emailSlide, trySlide]
    }
    
    fileprivate var currentSlideIndex: Int {
        guard scrollView.bounds != .zero else {
            return 0
        }
        
        return Int(ceil(scrollView.contentOffset.x / scrollView.bounds.size.width))
    }
    
    fileprivate var currentSlide: TutorialBaseSlideView {
        return slides[currentSlideIndex]
    }
    
    // MARK: - Initialization
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        
        slides = buildSlides()
        updatePromptHandler = UpdatePromptHandler(containerViewController: self)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // MARK: - UIViewController
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        track("Started Tutorial")
        
        updatePromptHandler.start()
        
        view.backgroundColor = .white
        
        // Setup scroll view
        scrollView.translatesAutoresizingMaskIntoConstraints = false
        scrollView.delegate = self
        scrollView.showsHorizontalScrollIndicator = false
        scrollView.isPagingEnabled = true
        scrollView.isScrollEnabled = false
        view.addSubview(scrollView)
        
        NSLayoutConstraint.activate([
            scrollView.topAnchor.constraint(equalTo: topLayoutGuide.bottomAnchor),
            scrollView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            scrollView.bottomAnchor.constraint(equalTo: bottomLayoutGuide.topAnchor),
            scrollView.trailingAnchor.constraint(equalTo: view.trailingAnchor)
        ])
        
        // Setup content view
        contentView.translatesAutoresizingMaskIntoConstraints = false
        scrollView.addSubview(contentView)
        
        NSLayoutConstraint.activate([
            contentView.topAnchor.constraint(equalTo: topLayoutGuide.bottomAnchor),
            contentView.leadingAnchor.constraint(equalTo: scrollView.leadingAnchor),
            contentView.bottomAnchor.constraint(equalTo: bottomLayoutGuide.topAnchor),
            contentView.trailingAnchor.constraint(equalTo: scrollView.trailingAnchor),
            contentView.widthAnchor.constraint(equalTo: view.widthAnchor, multiplier: CGFloat(slides.count)),
            contentView.heightAnchor.constraint(equalTo: scrollView.heightAnchor)
        ])
        
        prepareSlideViews()
    }
    
    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransition(to: size, with: coordinator)
        
        let currentSlideIdx = currentSlideIndex
        coordinator.animate(alongsideTransition: { context in
            var offset = self.scrollView.contentOffset
            offset.x = size.width * CGFloat(currentSlideIdx)
            self.scrollView.contentOffset = offset
        }, completion: nil)
    }
    
    // MARK: - Public
    
    func presentVideo() {
        let vc = TutorialVideoViewController(video: video)
        vc.modalTransitionStyle = .crossDissolve
        vc.delegate = self
        present(vc, animated: true, completion: nil)
    }
    
    // MARK: - Private
    
    fileprivate func scrollToNextSlide(animated: Bool = true) {
        guard scrollViewIsScrollingAnimation == false else {
            return
        }
        
        scrollViewIsScrollingAnimation = true
        currentSlide.willLeaveSlide()
        
        var offset = CGPoint.zero
        offset.x = scrollView.bounds.size.width + scrollView.contentOffset.x
        scrollView.setContentOffset(offset, animated: animated)
    }
    
    private func prepareSlideViews() {
        let padding = Geometry.padding()
        let top = contentView.layoutMargins.top
        
        slides.enumerated().forEach { i, slide in
            slide.translatesAutoresizingMaskIntoConstraints = false
            contentView.addSubview(slide)
            
            slide.layoutMargins = UIEdgeInsets(top: padding, left: padding, bottom: padding, right: padding)
            
            NSLayoutConstraint.activate([
                slide.widthAnchor.constraint(equalTo: scrollView.widthAnchor),
                slide.heightAnchor.constraint(equalTo: scrollView.heightAnchor, constant: -top),
                slide.topAnchor.constraint(equalTo: contentView.topAnchor, constant: top)
                ])
            
            if i == 0 {
                slide.leadingAnchor.constraint(equalTo: contentView.leadingAnchor).isActive = true
                slide.didEnterSlide()
                
            } else if i == slides.count - 1 {
                slide.trailingAnchor.constraint(equalTo: contentView.trailingAnchor).isActive = true
            }
            
            if i > 0 {
                let previousSlide = slides[i - 1]
                slide.leadingAnchor.constraint(equalTo: previousSlide.trailingAnchor).isActive = true
            }
        }
    }
    
    @objc fileprivate func dismissViewController() {
        dismiss(animated: true, completion: nil)
    }
}

extension TutorialViewController : UIScrollViewDelegate {
    func scrollViewDidEndScrollingAnimation(_ scrollView: UIScrollView) {
        scrollViewIsScrollingAnimation = false
        
        currentSlide.didEnterSlide()
    }
}

extension TutorialViewController : TutorialVideoViewControllerDelegate {
    func tutorialVideoViewControllerDoneButtonTapped(_ viewController: TutorialVideoViewController) {
        dismissViewController()
        scrollToNextSlide()
    }
}

extension TutorialViewController : TutorialEmailSlideViewDelegate {
    func tutorialEmailSlideViewDidComplete(_ slideView: TutorialEmailSlideView!) {
        slideView.delegate = nil
        scrollToNextSlide()
    }
    
    func tutorialEmailSlideViewDidTapPrivacyPolicy(_ slideView: TutorialEmailSlideView!) {
        if let vc = TutorialEmailSlideView.privacyPolicyViewController(withDoneTarget: self, doneAction: #selector(dismissViewController)) {
            present(vc, animated: true, completion: nil)
        }
    }
    
    func tutorialEmailSlideViewDidTapTerms(ofService slideView: TutorialEmailSlideView!) {
        if let vc = TutorialEmailSlideView.termsOfServiceViewController(withDoneTarget: self, doneAction:#selector(dismissViewController)) {
            present(vc, animated: true, completion: nil)
        }
    }
}

extension TutorialViewController : TutorialTrySlideViewDelegate {
    func tutorialTrySlideViewDidComplete(_ slideView: TutorialTrySlideView!) {
        slideView.delegate = nil
        
        UserDefaults.standard.set(true, forKey: UserDefaultsKeys.tutorialCompleted)
        track("Finished Tutorial")
        
        delegate?.tutoriaViewControllerDidComplete(self)
    }
}
